JavaClient stream response body using byte arrays

alex-james-dev opened the pull request:
- Currently, `JavaClient` streams the HTTP response body one byte at a time; this is slow. Make `JavaClient` stream the HTTP response body using byte arrays; this is faster. Each byte array can contain up to 4096 bytes.

- If none of the HTTP response body bytes are available without blocking, let the worker isolate's event loop run.

- Generate bindings for `java.io.BufferedInputStream` using `jnigen`.

- Create a `BufferedInputStream` from the HTTP response body `InputStream` to improve performance.

## Context
This is a PR for the `java_http` [GSoC '23 project](https://summerofcode.withgoogle.com/programs/2023/projects/NKUQqmSA) 🌞.
Relevant tracking issue: https://github.com/dart-lang/http/issues/957.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

brianquinlan commented on pkgs/java_http/lib/src/java_client.dart:
```diff
@@ -241,13 +242,24 @@ class JavaClient extends BaseClient {
+    // TODO: bufferedInputStream.read() could throw IOException.
+    while ((bytesRead = bufferedInputStream.read1(buffer, 0, buffer.length)) !=
+        -1) {
+      if (bytesRead == 0) continue;
```
```dart
if (bytesRead == 0) {
  // No more data is available without blocking so give other Isolates an opportunity
  // to run.
  await Future.delayed(Duration(seconds: 0));
}

alex-james-dev replied:
:heavy_check_mark: Done!

Thanks :smile:

alex-james-dev commented on pkgs/java_http/lib/src/java_client.dart:
```diff
@@ -230,24 +231,39 @@ class JavaClient extends BaseClient {
         ? httpUrlConnection.getInputStream()
         : httpUrlConnection.getErrorStream();
+    final bufferedInputStream = BufferedInputStream(inputStream);
+    print(bufferedInputStream.available());
```
I'll delete this.

alex-james-dev replied:
:heavy_check_mark: Done!

alex-james-dev commented on pkgs/java_http/lib/src/java_client.dart:
```diff
@@ -230,24 +231,39 @@ class JavaClient extends BaseClient {
-      actualBodyLength++;
+    final buffer = JArray(jbyte.type, 4096);
+    // TODO: bufferedInputStream.read() could throw IOException.
+    while ((bytesRead = bufferedInputStream.read1(buffer, 0, buffer.length)) !=
+        -1) {
```
I'm not too keen on the while loop condition going over to the next line, I think I will refactor it to make the code cleaner.

alex-james-dev replied:
:heavy_check_mark: Done!

To simplify the while loop condition, we no longer do the variable assignment within the loop condition.

alex-james-dev commented:
I renamed the PR title to be <= 50 characters :smile:

alex-james-dev commented:
I've done a bit of refactoring to try to clean the code up a bit:

- Renamed some variables.

- Rewrote the while loop condition so that it fits on one line.

- Created an extension method for converting from `JArray<jbyte>` to `Uint8List`.

The PR is now ready for review! :tada: 

Thank you!

brianquinlan commented on pkgs/java_http/lib/src/java_client.dart:
```diff
@@ -230,35 +231,58 @@ class JavaClient extends BaseClient {
+    var bytesCount =
+        responseBodyStream.read1(bytesBuffer, 0, bytesBuffer.length);
+
+    while (bytesCount != -1) {
```
This might be easier:

```dart
while (true) {
  final bytesCount = responseBodyStream.read1(bytesBuffer, 0, bytesBuffer.length); 
  if (bytesCount == -1) {
    break;
  }
      if (bytesCount == 0) {
        // No more data is available without blocking so give other Isolates an
        // opportunity to run.
        await Future<void>.delayed(Duration.zero);
        continue;
    }

   sendPort.send(bytesBuffer.toUint8List(length: bytesCount));
   actualBodyLength += bytesCount;
}

}

alex-james-dev replied:
Hey @brianquinlan,

I like your approach because it removes the duplication of:

```dart 
bytesCount = responseBodyStream.read1(bytesBuffer, 0, bytesBuffer.length);
```

Including a continue statement when bytesCount == 0 is a good idea.

I've pushed both of these changes :muscle:. 

Thank you!
